File: src/aoc/2022/day-8/main.py
from pathlib import Path

def create_matrix(size: int):
  return [
    [0 for i in range(size)] 
    for index in range(size)
  ]

def create_tree_grid(lines: list[str], size: int):
  grid = create_matrix(size)
  for row_index, line in enumerate(lines):
    for column_index, char in enumerate(line):
      grid[row_index][column_index] = int(char)
  return grid

def get_file_lines(file_path: str):
  file = Path(__file__).with_name(file_path)
  return file.read_text().splitlines()

def get_visible_tree_count(lines: list[str], size: int):
  count = 0
  grid = create_tree_grid(lines, size)
  for row_index, row in enumerate(grid):
    for column_index, tree_height in enumerate(row):
      if is_visible(grid, row_index, column_index, tree_height):
        count += 1
  return count

def is_in_edge(grid: list[list[int]], r_idx: int, c_idx: int):
  last_element_index = len(grid) - 1
  return (
    r_idx == 0 or r_idx == last_element_index or
    c_idx == 0 or c_idx == last_element_index
  )

def is_visible(grid: list[list[int]], row_index: int, column_index: int, tree_height: int):
  return (
    is_in_edge(grid, row_index, column_index) or
    is_visible_from(grid, row_index, column_index, tree_height, "top") or 
    is_visible_from(grid, row_index, column_index, tree_height, "right") or
    is_visible_from(grid, row_index, column_index, tree_height, "bottom") or
    is_visible_from(grid, row_index, column_index, tree_height, "left")  
  )

def is_visible_from(grid: list[list[int]], r_idx: int, c_idx: int, tree_height: int, direction: str):
  length = len(grid)
  if direction == "top":
    for row_index in range(r_idx - 1, -1, -1):
      height = grid[row_index][c_idx]
      if height >= tree_height:
        return False 
  elif direction == "right":
    for column_index in range(c_idx + 1, length):
      height = grid[r_idx][column_index]
      if height >= tree_height:
        return False
  elif direction == "bottom":
    for row_index in range(r_idx + 1, length):
      height = grid[row_index][c_idx]
      if height >= tree_height:
        return False
  elif direction == "left":
    for column_index in range(c_idx - 1, -1, -1):
      height = grid[r_idx][column_index]
      if height >= tree_height:
        return False
  else:
    raise ValueError(f"{direction} is an invalid direction")
  return True

def print_results(file_name: str, count: int):
  print(f"{file_name} File:")
  print(f"Visible Trees Count: {count}")

FILE_PATH_EXAMPLE = "./example.txt"
FILE_PATH_INPUT = "./input.txt"
GRID_SIZE_EXAMPLE = 5
GRID_SIZE_INPUT = 99

lines = get_file_lines(FILE_PATH_EXAMPLE)
count = get_visible_tree_count(lines, GRID_SIZE_EXAMPLE)
print_results("Example", count)

lines = get_file_lines(FILE_PATH_INPUT)
count = get_visible_tree_count(lines, GRID_SIZE_INPUT)
print_results("Example", count)

File: src/aoc/2022/day-8/main.rb
def create_matrix size
  Array.new(size) { Array.new(size) }
end 

def create_tree_grid lines, size
  grid = create_matrix(size)
  lines.each.with_index do |line, row_index|
    line.each_char.with_index do |char, column_index|
      grid[row_index][column_index] = char.to_i
    end
  end
  return grid
end

def get_file_lines file_path
  IO.readlines(file_path, chomp: true)
end

def get_max_tree_scenic_score lines, size
  max = 0
  grid = create_tree_grid(lines, size)
  grid.each.with_index do |row, row_index|
    row.each.with_index do |tree_height, column_index|
      scenic_score = get_tree_scenic_score(grid, row_index, column_index, tree_height)
      max = scenic_score if scenic_score > max
    end
  end
  return max
end


def get_tree_scenic_score grid, row_index, column_index, tree_height
  scores = DIRECTIONS.map do |direction|
    count = 0
    use_direction_loop(grid, row_index, column_index, tree_height, direction, count) do |height|
      count += 1
      if tree_height <= height
        true 
      end
    end
    count
  end
  return scores.reduce(:*)
end
  
def get_visible_tree_count lines, size
  count = 0
  grid = create_tree_grid(lines, size)
  grid.each.with_index do |row, row_index|
    row.each.with_index do |tree_height, column_index|
      count += 1 if is_visible(grid, row_index, column_index, tree_height)
    end
  end
  return count
end

def is_in_edge grid, row_index, column_index
  last_element_index = grid.length - 1
  in_row_edge = row_index == 0 || row_index == last_element_index || 
  in_column_edge = column_index == 0 || column_index == last_element_index
  return in_row_edge || in_column_edge
end

def is_visible grid, row_index, column_index, tree_height
  is_in_edge(grid, row_index, column_index) ||
  is_visible_from(grid, row_index, column_index, tree_height, "top") ||
  is_visible_from(grid, row_index, column_index, tree_height, "right") ||
  is_visible_from(grid, row_index, column_index, tree_height, "bottom") ||
  is_visible_from(grid, row_index, column_index, tree_height, "left")
end 

def is_visible_from grid, row_index, column_index, tree_height, direction
  use_direction_loop(grid, row_index, column_index, tree_height, direction, true) do |height|
    if height >= tree_height
      [true, false]
    end
  end 
end

def print_results file_name, count, scenic_score
  puts "#{file_name} File Results:"
  puts "Visible Trees Count: #{count}"
  puts "Maximum Tree Scenic Score: #{scenic_score}"
end

def use_direction_loop grid, row_index, column_index, tree_height, direction, initial_value
  len = grid.length
  case direction
  when "top"
    (row_index - 1).downto(0) do |r_idx|
      height = grid[r_idx][column_index]
      return_early, value = yield height
      return value if return_early
    end
  when "right"
    (column_index + 1).upto(len - 1) do |c_idx|
      height = grid[row_index][c_idx]
      return_early, value = yield height
      return value if return_early
    end
  when "bottom"
    (row_index + 1).upto(len - 1) do |r_idx|
      height = grid[r_idx][column_index]
      return_early, value = yield height
      return value if return_early
    end
  when "left"
    (column_index - 1).downto(0) do |c_idx|
      height = grid[row_index][c_idx]
      return_early, value = yield height
      return value if return_early
    end
  end
  return initial_value
end

DIRECTIONS = ["top", "right", "bottom", "left"]
FILE_PATH_EXAMPLE = "./example.txt"
FILE_PATH_INPUT = "./input.txt"
GRID_SIZE_EXAMPLE = 5
GRID_SIZE_INPUT = 99

lines = get_file_lines(FILE_PATH_EXAMPLE)
count = get_visible_tree_count(lines, GRID_SIZE_EXAMPLE)
scenic_score = get_max_tree_scenic_score(lines, GRID_SIZE_EXAMPLE)
print_results("Example", count, scenic_score)

lines = get_file_lines(FILE_PATH_INPUT)
count = get_visible_tree_count(lines, GRID_SIZE_INPUT)
scenic_score = get_max_tree_scenic_score(lines, GRID_SIZE_INPUT)
print_results("Input", count, scenic_score)
